File: src/main/org/nlogo/prim/etc/_wrapcolor.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim.etc;

import org.nlogo.api.Syntax;
import org.nlogo.nvm.Context;
import org.nlogo.nvm.Pure;
import org.nlogo.nvm.Reporter;

public final strictfp class _wrapcolor extends Reporter implements Pure {
  @Override
  public Syntax syntax() {
    return Syntax.reporterSyntax
        (new int[]{Syntax.NumberType()},
            Syntax.NumberType());
  }

  @Override
  public Object report(Context context) {
    return report_1(context, argEvalDoubleValue(context, 0));
  }

  public double report_1(Context context, double colorValue) {
    if (colorValue < 0 || colorValue >= 140) {
      return org.nlogo.api.Color.modulateDouble(colorValue);
    } else {
      return colorValue;
    }
  }
}

File: src/main/jvm/org/nlogo/prim/etc/_wrapcolor.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim.etc

import org.nlogo.api.{ Syntax, Color }
import org.nlogo.nvm.{ Context, Pure, Reporter }

class _wrapcolor extends Reporter with Pure {

  override def syntax =
    Syntax.reporterSyntax(
      Array(Syntax.NumberType), Syntax.NumberType)

  override def report(context: Context) =
    Double.box(report_1(context, argEvalDoubleValue(context, 0)))

  def report_1(context: Context, colorValue: Double): Double =
    if (colorValue < 0 || colorValue >= 140)
      Color.modulateDouble(colorValue)
    else
      colorValue

}
